### ISSUE TITLE
EulerNet 效果异常问题

### ISSUE THREAD

**Issue by Ethan-TZ** [salmon1802/DCNv3#2]
您好，非常感谢您对EulerNet的关注。然而，我们发现你的paper中Criteo数据集EulerNet的效果存在异常，可能是实现错误或者没调参。另外其它数据集的效果也很奇怪。麻烦请按照 https://github.com/reczoo/FuxiCTR/pull/107 的配置进行实验，谢谢。

**Comment by salmon1802** [comment]
感谢您对我们工作的关注 :)，我们对于EulerNet的实现使用的是您[https://github.com/Ethan-TZ/EulerNet/blob/main/Model/EulerNet.py]中的源代码。我们注意到造成当前性能评估差异的原因可能是"shape": "[20, 20]"这个关键参数没有进行尝试，我们之前设置的超参数是根据您论文中的Euler interaction
layer={1,2,3,4,5}与阶向量的个数=39进行设置并利用网格搜索进行性能调优的。我们将在论文的更新版本中，利用您开源的超参数配置对实验结果进行进一步的修正。祝好。

**Comment by salmon1802** [comment]
您好，我在[reczoo/FuxiCTR#107](https://github.com/reczoo/FuxiCTR/pull/107)中观察到，EulerNet在Criteo数据集上的运行日志中显示总参数量为16602400，而我按照您在[reczoo/FuxiCTR#107](https://github.com/reczoo/FuxiCTR/pull/107)中的配置进行实验后发现，打印出的总参数量为14878016。以下是我的运行日志，想请教您一下可能得原因。

2024-07-29 16:06:31,998 P1010284 INFO Params: {
    "batch_norm": "True",
    "batch_size": "10000",
    "data_format": "h5",
    "data_root": "../../data/",
    "dataset_id": "Criteo_x4_10_h5",
    "debug_mode": "False",
    "early_stop_patience": "2",
    "embedding_dim": "16",
    "embedding_regularizer": "1e-05",
    "epochs": "100",
    "eval_steps": "None",
    "feature_cols": "[{'active': True, 'dtype': 'float', 'fill_na': 0, 'na_value': 0, 'name': ['I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9', 'I10', 'I11', 'I12', 'I13'], 'preprocess': 'convert_to_bucket', 'type': 'categorical'}, {'active': True, 'dtype': 'str', 'fill_na': '', 'na_value': '', 'name': ['C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'C8', 'C9', 'C10', 'C11', 'C12', 'C13', 'C14', 'C15', 'C16', 'C17', 'C18', 'C19', 'C20', 'C21', 'C22', 'C23', 'C24', 'C25', 'C26'], 'type': 'categorical'}]",
    "feature_config": "None",
    "feature_specs": "None",
    "gpu": "0",
    "group_id": "None",
    "label_col": "{'dtype': 'float', 'name': 'Label'}",
    "layer_norm": "True",
    "learning_rate": "0.001",
    "loss": "binary_crossentropy",
    "metrics": "['logloss', 'AUC']",
    "min_categr_count": "10",
    "model": "EulerNet",
    "model_id": "EulerNet_Criteo",
    "model_root": "./checkpoints/",
    "monitor": "{'AUC': 1, 'logloss': -1}",
    "monitor_mode": "max",
    "net_ex_dropout": "0.1",
    "net_im_dropout": "0.1",
    "net_regularizer": "0",
    "num_workers": "3",
    "optimizer": "adam",
    "pickle_feature_encoder": "True",
    "save_best_only": "True",
    "seed": "2021",
    "shape": "[20, 20]",
    "shuffle": "True",
    "task": "binary_classification",
    "test_data": "../../data/Criteo_x4_10_h5/test.h5",
    "train_data": "../../data/Criteo_x4_10_h5/train.h5",
    "use_features": "None",
    "valid_data": "../../data/Criteo_x4_10_h5/valid.h5",
    "verbose": "1"
}
2024-07-29 16:06:31,999 P1010284 INFO Load feature_map from json: ../../data/Criteo_x4_10_h5/feature_map.json
2024-07-29 16:06:31,999 P1010284 INFO Set column index...
2024-07-29 16:06:31,999 P1010284 INFO Feature specs: {
    "C1": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 1445, 'vocab_size': 1446}",
    "C10": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 39529, 'vocab_size': 39530}",
    "C11": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 5130, 'vocab_size': 5131}",
    "C12": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 156655, 'vocab_size': 156656}",
    "C13": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 3175, 'vocab_size': 3176}",
    "C14": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 27, 'vocab_size': 28}",
    "C15": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 11042, 'vocab_size': 11043}",
    "C16": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 148912, 'vocab_size': 148913}",
    "C17": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 11, 'vocab_size': 12}",
    "C18": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 4559, 'vocab_size': 4560}",
    "C19": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 2002, 'vocab_size': 2003}",
    "C2": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 553, 'vocab_size': 554}",
    "C20": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 4, 'vocab_size': 5}",
    "C21": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 154563, 'vocab_size': 154564}",
    "C22": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 17, 'vocab_size': 18}",
    "C23": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 16, 'vocab_size': 17}",
    "C24": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 53030, 'vocab_size': 53031}",
    "C25": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 81, 'vocab_size': 82}",
    "C26": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 40954, 'vocab_size': 40955}",
    "C3": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 157338, 'vocab_size': 157339}",
    "C4": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 117821, 'vocab_size': 117822}",
    "C5": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 305, 'vocab_size': 306}",
    "C6": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 17, 'vocab_size': 18}",
    "C7": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 11881, 'vocab_size': 11882}",
    "C8": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 629, 'vocab_size': 630}",
    "C9": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 4, 'vocab_size': 5}",
    "I1": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 43, 'vocab_size': 44}",
    "I10": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 5, 'vocab_size': 6}",
    "I11": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 26, 'vocab_size': 27}",
    "I12": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 36, 'vocab_size': 37}",
    "I13": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 71, 'vocab_size': 72}",
    "I2": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 98, 'vocab_size': 99}",
    "I3": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 121, 'vocab_size': 122}",
    "I4": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 40, 'vocab_size': 41}",
    "I5": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 219, 'vocab_size': 220}",
    "I6": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 111, 'vocab_size': 112}",
    "I7": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 79, 'vocab_size': 80}",
    "I8": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 68, 'vocab_size': 69}",
    "I9": "{'source': '', 'type': 'categorical', 'padding_idx': 0, 'oov_idx': 91, 'vocab_size': 92}"
}
2024-07-29 16:06:35,385 P1010284 INFO Total number of parameters: 14878016.
2024-07-29 16:06:35,385 P1010284 INFO Loading data...
2024-07-29 16:06:35,386 P1010284 INFO Loading data from h5: ../../data/Criteo_x4_10_h5/train.h5
2024-07-29 16:07:20,676 P1010284 INFO Train samples: total/36672493, blocks/1
2024-07-29 16:07:20,677 P1010284 INFO Loading data from h5: ../../data/Criteo_x4_10_h5/valid.h5
2024-07-29 16:07:23,960 P1010284 INFO Validation samples: total/4584062, blocks/1
2024-07-29 16:07:23,960 P1010284 INFO Loading train and validation data done.
2024-07-29 16:07:23,960 P1010284 INFO Start training: 3668 batches/epoch
2024-07-29 16:07:23,960 P1010284 INFO ************ Epoch=1 start ************
 34%|█████████████▎                         | 1254/3668 [01:12<02:14, 18.00it/s]

**Comment by salmon1802** [comment]
实验结果与对应论文部分已更新
